src/bookshelf/src/bksw_sec.erl

```
@@ -35,109 +37,283 @@ is_authorized(Req0, #context{} = Context) ->
-                false ->
-                    case ((erlang:iolist_to_binary(AWSAccessKeyId) ==
-                               erlang:iolist_to_binary(AccessKey)) andalso
-                          erlang:iolist_to_binary(Signature) ==
-                              erlang:iolist_to_binary(IncomingSignature)) of
-                        true ->
-                            MaxAge = "max-age=" ++ integer_to_list(ExpireDiff),
-                            Req1 = wrq:set_resp_header("Cache-Control", MaxAge, Req0),
-                            {true, Req1, Context};
-                        false ->
-                            ?LOG_DEBUG("req_id=~p signing error for ~p", [ReqId, Path]),
-                            encode_sign_error_response(AWSAccessKeyId, IncomingSignature, RequestId,
-                                                       StringToSign, Req0, Context)
-                    end
-                end;
-        error ->
-            encode_access_denied_error_response(RequestId, Req0, Context)
+% presigned url verification
+% https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-query-string-auth.html
+presigned_auth(RequestId, Req0, Context, Headers0) ->
+    Credential = wrq:get_qs_value("X-Amz-Credential", "", Req0),
+    XAmzDate = wrq:get_qs_value("X-Amz-Date", "", Req0),
+    SignedHeaderKeysString = wrq:get_qs_value("X-Amz-SignedHeaders", "", Req0),
+    IncomingSignature = wrq:get_qs_value("X-Amz-Signature", "", Req0),
+    % 1 =< XAmzExpires =< 604800
+    XAmzExpiresString = wrq:get_qs_value("X-Amz-Expires", "", Req0),
+    common_auth(RequestId, Req0, Context, Credential, XAmzDate, SignedHeaderKeysString, IncomingSignature, XAmzExpiresString, Headers0, presigned_url).
+
+% https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-auth-using-authorization-header.html
+header_auth(RequestId, IncomingAuth, Req0, Context, Headers0) ->
+    try
+        (ParseAuth = parse_authorization(IncomingAuth)) /= err orelse throw({RequestId, Req0, Context}),
```

I'm still getting my bearings with this change, but given the importance of this code, I think we want to avoid `throw/catch` style error handling here and prefer explicit returns with error handling.  Happy to chat about it though if I'm off base here.

---

Nothing's off base :)  I'm flexible and open-minded.  Throw me an example snippet for what you have in mind and I'll get to work.

---

Ah, I see 😃  The reasons I did it the way I did was 1) to avoid long, nested case statements, 2) bail early on error conditions, and 3) I was having trouble with errors showing up in logs with a "let it crash" style (a style which I prefer, btw).  Having said all of that, I'm happy to rewrite this and see where we stand after.